python/lsst/pipe/base/shims.py

```diff
@@ -0,0 +1,320 @@
+            parameters = dict(bbox=fullDataId.pop("bbox"))
+            origin = fullDataId.pop("origin", lsst.afw.image.PARENT)
+            parameters["origin"] = origin
+        else:
+            parameters = {}
+        try:
+            return self._butler3.get(datasetType, fullDataId, parameters=parameters)
+        except (FileNotFoundError, LookupError) as err:
+            raise NoResults(str(err), datasetType, fullDataId)
+
+    def put(self, obj, datasetType, dataId=None, doBackup=False, **rest):
+        """Write a dataset.
+
+        Parameters
+        ----------
+        obj
+            Object to write.
+        datasetType : `str`
+            Name of the Gen2 dataset type.
+        dataId : `dict` or `~lsst.daf.butler.DataId`, optional
+            A Generation 3 data ID that identifies the dataset.
+        doBackup : `bool`
+            This option is provided for compatibility with
+            `lsst.daf.persistence.Butler`, but must be `False`.
+        rest
+            Additional key-value pairs to augment the given data ID.
+        """
+        if doBackup:
+            raise NotImplementedError("ShimButler cannot implement put with 'doBackup=True'")
+        datasetType = self._translateDatasetType(datasetType)
+        self._butler3.put(obj, datasetType, self._makeDataId(dataId, **rest))
+
+    def dataRef(self, datasetType, level=None, dataId=None, **rest):
+        """Return a DataRef associated with the given dataset type and data ID.
+
+        Parameters
+        ----------
+        datasetType : `str`
+            Name of the dataset type.
+        dataId : `dict` or `~lsst.daf.butler.DataId`, optional
+            A Generation 3 data ID that identifies the dataset.
+        level
+            This option is provided for compatibility with
+            `lsst.daf.persistence.Butler`, but must be `None`.
+        rest
+            Additional key-value pairs to augment the given data ID.
+        """
+        if level is not None:
+            raise NotImplementedError("ShimButler cannot implement dataRef with 'level != None'")
+        fullDataId = {}
+        if dataId is not None:
+            fullDataId.update(dataId)
+        fullDataId.update(rest)
+        return next(iter(ShimButlerSubset(self, datasetType, [fullDataId])))
+
+
+class ShimButlerSubset:
+    """A shim for a Gen2 `~lsst.daf.persistence.ButlerSubset` with a Gen3
+    `~lsst.daf.butler.Butler` backend.
+
+    Parameters
+    ----------
+    butler : `ShimButler`
+        Butler shim instance.
+    datasetType : `str`
+        Name of the dataset type.
+    dataIds : iterable of `dict` or `~lsst.daf.butler.DataId`
+        Generation 3 data IDs that define the data in this subset.
+    """
+
+    GENERATION = 2.5
+    """This is a Generation 2 shim for a Generation3 Butler.
+    """
+
+    def __init__(self, butler, datasetType, dataIds):
+        self.butler = butler
+        self.datasetType = datasetType
+        self._dataIds = list(dataIds)
```

Comment:
If you are going to have iter defined in the same class as it is iterating over, then I think that _dataIds should be constructed as a tuple instead of a list. This will prevent someone from messing with the _dataIds while an existing iterator is outstanding. Unless of course this is needed for backwards compatibility were people abuse modifying the private variable. The other option is to have __iter__ construct and return an object that has its own list pointing to each of the elements.